demo-circuit/src/naive_threshold_sig_w_key_rotation/mod.rs

```diff
@@ -0,0 +1,789 @@
+        // Enforce H(validator key root, threshold) to be equal to H(genesis_validator_keys_tree_root_g, threshold)
+        // or H(all_leaves_root_g, threshold) if it is the first certificate
+
+        //in the case of the first certificate, check if expected_vkroot_threshold_hash_g equals to evaluated vkroot_threshold_hash_g with initial set of keys
+        //otherwise, in case of new certificates, use the genesis_validator_keys_tree_root
+        let mut vkroot_threshold_hash_g = FieldElementGadget::conditionally_select(
+            cs.ns(|| "select initial keys for threshold check"), 
+            &is_first_certificate_g, 
+            &validator_mktree_leaves_root_g, 
+            &genesis_validator_keys_tree_root_g,
+        )?;
+
+        vkroot_threshold_hash_g = FieldHashGadget::enforce_hash_constant_length(
+            cs.ns(|| "H(H(pks), threshold)"),
+            &[vkroot_threshold_hash_g, threshold_g.clone()],
+        )?;
+
+        let expected_vkroot_threshold_hash_g =
+            FieldElementGadget::alloc_input(cs.ns(|| "alloc constant as input"), || {
+                self.genesis_constant
+                    .ok_or(SynthesisError::Other("AssignmentMissing genesis constant".to_string()))
+            })?;
+
+            vkroot_threshold_hash_g.enforce_equal(
+            cs.ns(|| "vkroot_threshold_hash: expected == actual"),
+            &expected_vkroot_threshold_hash_g,
+        )?;
+
+
+        // ****************** 3 Verify wcert signatures ********************
+        // for (i <- 0 to (VALIDATORS_SIZE-1)) {
+        //     var sigNum = 0                  // we will count number of collected signatures
+        //     if (wcert_signatures[i] != NULL) {
+        //         val is_valid_sig = verify_signature(sc_wcert_hash, wcert_signatures[i], signing_keys[i])      // order of signatures should corresponed the order of keys in signing_keys
+        //         require(is_valid_sig == true)
+        //         sigNum = sigNum + 1    
+        //     }
+        // }
+        // require(sigNum > VALIDATORS_SIZE/2)  CR: This is replaced by the threshold value
+   
+        //Alloc the b's as witnesses
+        let mut bs_g = Vec::with_capacity(log_max_pks + 1);
+        for (i, b) in self.b.iter().enumerate() {
+            let b_g = Boolean::alloc(cs.ns(|| format!("alloc b_{}", i)), || {
+                b.ok_or(SynthesisError::Other(format!("AssignmentMissing b {}", i)))
+            })?;
+            bs_g.push(b_g);
+        }
+
+        //Pack the b's into a field element
+        let b_field = FieldElementGadget::from_bits(
+            cs.ns(|| "pack the b's into a field element"),
+            bs_g.as_slice(),
+        )?;
+
+        // 4 ************** Verify quality (We assume `quality` is the number of signatures in `signatures` vector)  *************
+        // require(quality == sigNum)
+
+        //Enforce threshold
+        valid_signatures
```

Comment:
Since we refactor to use the WithdrawalCertificate data struct, already containing the quality, we need to enforce valid_signature == current_wcert.quality

Reply:
I think its already there in the circuit, see [here](https://github.com/HorizenOfficial/zendoo-sc-cryptolib/blob/ml/key_rotation/demo-circuit/src/naive_threshold_sig_w_key_rotation/mod.rs#L673)

Reply:
This check just enforces that the number of valid signatures is above the threshold. We need to enforce them to be equal to the quality present in the current wcert. Actually, by looking at the new function, I'm afraid that we are not replacing the quality field (that is set to 0 Java side) to be equal to the number of valid signatures (that we are supposed to compute before passing the certificate to the circuit)

Reply:
Are we wanting to calculate the quality here or have Java do it and pass it then check that they're equal?

Reply:
So my idea is:
1. Like you are doing now, Java side they will construct a certificate without quality, that internally means setting `quality = 0`;
2. Eventually, the `NaiveThresholdSignatureWKeyRotation::new` function is called: in this function we compute the msg_to_sign, we compute the number of valid signatures, and we set `current_wcert.quality = num_valid_sigs`. After that we can compute the current certificate hash and construct the circuit instance;
3. In the circuit, among the other checks, we will do: `valid_signatures.enforce_equal(.., current_wcert_g.quality)`

Reply:
Done